src/Game.sol

```diff
@@ -0,0 +1,142 @@
+    mapping(uint256 => mapping(uint256 => uint256)) public Hex;
+
+    /// @notice For each environment, lists the cumulative distribution function for terrains in that environment.
+    /// @dev Environments are: 0 - forest, 1 - prairie, 2 - river, 3 - arctic, 4 - marsh, 5 - badlands, 6 - hills
+    /// @dev Terrain types are: 0 - plain, 1 - forest, 2 - swamp, 3 - water, 4 - mountain, 5 - desert, 6 - ice
+    uint8[7][7] public EnvironmentDistributions = [
+        [0, 90, 98, 123, 128, 128, 128],
+        [90, 95, 100, 120, 120, 128, 128],
+        [0, 0, 8, 128, 128, 128, 128],
+        [0, 8, 8, 8, 28, 28, 128],
+        [5, 10, 108, 128, 128, 128, 128],
+        [18, 18, 18, 18, 28, 128, 128],
+        [0, 43, 43, 48, 128, 128, 128]
+    ];
+
+    /// @param blocksToAct Number of blocks that a player has to decide whether to accept their fate or re-roll. This parameter applies to every such decision point.
+    /// @param rollFee Fee for first roll on any action.
+    /// @param rerollFee Fee for re-roll on any action, assuming player doesn't want to accept their fate.
+    constructor(uint256 blocksToAct, uint256 rollFee, uint256 rerollFee)
+        Bandit(blocksToAct, address(this), rollFee, rerollFee)
+        ERC20("Supply", "SUPPLY")
+    {
+        uint256 prevBlockNumber = 0;
+        if (block.number > 0) {
+            prevBlockNumber = block.number - 1;
+        }
+        // Zero out the leading 12 bits of the block hash, to prevent overflows when adding 31 * j.
+        uint256 startingEntropy = uint256(blockhash(prevBlockNumber)) >> 12 << 12;
+        for (uint256 j = 0; j < 100; j++) {
+            _explore(0, 2 * j, startingEntropy + (31 * j));
+        }
+    }
+
+    /// @notice Internal method that explores a hex and sets its state.
+    function _explore(uint256 i, uint256 j, uint256 entropy) internal {
+        uint256 env = environment(i);
+        uint8 maskedEntropy = uint8(entropy & u7mask);
+        if (maskedEntropy < EnvironmentDistributions[env][6]) {
+            // 1111
+            Hex[i][j] = 13;
+        } else if (maskedEntropy < EnvironmentDistributions[env][5]) {
+            // 1011
+            Hex[i][j] = 11;
+        } else if (maskedEntropy < EnvironmentDistributions[env][4]) {
+            // 1001
+            Hex[i][j] = 9;
+        } else if (maskedEntropy < EnvironmentDistributions[env][3]) {
+            // 0111
+            Hex[i][j] = 7;
+        } else if (maskedEntropy < EnvironmentDistributions[env][2]) {
+            // 0101
+            Hex[i][j] = 5;
+        } else if (maskedEntropy < EnvironmentDistributions[env][1]) {
+            // 0011
+            Hex[i][j] = 3;
+        } else {
+            // 0001
+            Hex[i][j] = 1;
+        }
+    }
+
+    /// @notice Describes the environment of a hex with the given j-coordinate.
+    function environment(uint256 i) public pure returns (uint256) {
```

Comment:
I don't see any tests for the environment assignments. I don't know that we need them, but maybe a few comments about the flow of the environment and what this calculation is doing? Comment that the environment is constant for 32 column blocks?

Reply:
In Game.t.sol:

![image](https://github.com/moonstream-to/degen-trail/assets/8016073/bdec6726-312c-40cf-9c47-88062f02b62c)